near-sdk/src/store/index_map.rs

```diff
@@ -21,7 +21,7 @@ where
     ///
     /// Note: u32 indices are used over usize to have consistent functionality across architectures.
     /// Some functionality would be different from tests to Wasm if exceeding 32-bit length.
-    #[borsh_skip]
+    #[borsh(skip, bound(deserialize = ""))]
```

Comment:
@dj8yfo Should we fix it on borsh side? I feel `#[borsh(skip)]` should be enough and adding `bound()` here is an indicator of a deficiency in borsh.

Reply:
Here's a link to [this case in migration guide](https://github.com/near/borsh-rs/pull/201/files#diff-d94d69a1dd0393d69f6594f304958b8893aa8749dfbc32cb15cd4addddbdd046R348).

1. We've inherited/copied this behaviour from `serde`. 
2. `core::default::Default` for skipped field derivation
It's a logical thing to do in majority of cases: straightforward implementation of `Default` for `struct SomeStruct<T>`  will require `T: Default`, e.g., when one of fields contains `T` directly. 
In this case, `StableMap` is a collection, and it has a `Default` implementation, returning empty collection irrelevant of the values and bounds on type parameters.

Reply:
A similar example has aslo been published to [1.0.0-alpha.3/borsh doc](https://docs.rs/borsh/1.0.0-alpha.3/borsh/derive.BorshDeserialize.html#interaction-with-borshskip)

Reply:
By the way, the derived implementations `#[derive(BorshSerialize, BorshDeserialize)]` won't actually serialize/deserialize the collection, which was my first impression, the heavy lifting is done in `flush` and `get` methods, 
they will only serialize/deserialize the `prefix` of the collection.

Reply:
Actually i've discovered a [test](https://github.com/near/near-sdk-rs/blob/c2352df2491dcc979febfc4650c65bdba4bceb6c/near-sdk/src/store/vec/mod.rs#L885) of `Vector<T>`, based on `IndexMap<T>`, which tests exactly this behaviour of restoring collection just by restoring its `prefix`, so `#[derive(BorshSerialize, BorshDeserialize)]` isn't accidental, they were added by design.